### PR TITLE
fix: distinguish null group-by keys

### LINKED ISSUE #906: Group By merges NULL keys with valid 0/-1 values
## Problem

`GROUP BY` can merge a NULL key with a valid numeric key. The problem exists in both the optimized single-key path and the generic encoded-key path.

## Case 1: optimized single-key path

```cypher
UNWIND CAST([0, CAST(NULL, 'INT64'), CAST(NULL, 'INT64')], 'INT64[]') AS x
RETURN x, COUNT(*);
```

Current result:

```text
0 | 3
```

Expected result:

```text
0    | 1
NULL | 2
```

The optimized path reads the value returned by the expression directly. For nullable numeric expressions, the returned placeholder can be `0`, so NULL and the valid value `0` are inserted into the same hash-map entry.

## Case 2: generic encoded-key path

The second key below forces the generic multi-key path:

```cypher
UNWIND CAST([CAST(-1, 'INT32'), CAST(NULL, 'INT32')], 'INT32[]') AS x
UNWIND [7] AS y
RETURN x, y, COUNT(*);
```

Current result:

```text
-1 | 7 | 2
```

Expected result:

```text
-1   | 7 | 1
NULL | 7 | 1
```

`encode_value()` currently encodes NULL as an `int32_t` value of `-1`. Therefore, NULL and a valid `INT32 -1` produce the same encoded grouping key.

## Existing validity information

No new nullable key type is required for an initial fix. Both paths already expose validity information that can be applied directly before hashing or encoding:

- The optimized expression wrapper provides `expr.has_value(row)`.
- The generic path obtains a `Value`, which provides `Value::IsNull()`.

The optimized path can keep a separate NULL group, while the generic path can encode an explicit validity marker before each key component. This preserves the existing non-NULL key representation while making NULL distinct from every valid value.

### PR DESCRIPTION
**Pull request by Spockkk0225**
Fixes #906.

## Changes

- Keep a dedicated NULL group in the optimized single-key grouping path by checking `expr.has_value(row)`.
- Encode an explicit validity marker for each component in the generic grouping path by checking `Value::IsNull()`.
- Add regression tests covering NULL collisions with valid `INT64 0` and `INT32 -1` values.

## Tests

- Built the Python binding successfully.
- `test_empty_grouped_aggregate`
- `test_empty_ungrouped_count`
- `test_group_by_preserves_null_keys`

### CODE REVIEW

File: src/execution/execute/ops/retrieve/group_by_utils.cc [hunk 1]
```diff
@@ -42,7 +42,12 @@ struct GKey : public KeyBase {
       ::neug::Encoder encoder(buf);
       for (size_t k_i = 0; k_i < exprs.size(); ++k_i) {
         auto val = exprs[k_i]->get_elem(i);
-        encode_value(val, encoder);
+        if (val.IsNull()) {
```
Comment (liulx20) [inline src/execution/execute/ops/retrieve/group_by_utils.cc]: 建议在encoder里面处理null,可以直接append("NULL"),不建议用0填充

Comment (Spockkk0225) [reply src/execution/execute/ops/retrieve/group_by_utils.cc]: 现在按照要求改为显式的 `encoder.put_bytes("NULL", 4)` 和 `encoder.put_bytes("VALUE", 5)`。
关键在于null和非null都要加一个前缀用于区分，否则仍然会出现碰撞。

File: src/common/types/value.cc [hunk 2]
```diff
@@ -855,9 +855,10 @@ rapidjson::Value Value::ToJson(const Value& value,
 void encode_value(const Value& val, Encoder& encoder) {
   const auto& type = val.type();
   if (val.IsNull()) {
-    encoder.put_int(-1);
+    encoder.put_bytes("NULL", 4);
     return;
   }
+  encoder.put_bytes("VALUE", 5);
```
Comment (liulx20) [inline src/common/types/value.cc]: 这个是必须加的吗，这个函数影响面比较广，可能会有性能回归

**Comment by Spockkk0225** [comment]
改为在group by之前预留一个bitmap，遇到null时额外写入一个bit，其他逻辑和encoder的处理保持不变。

File: src/execution/execute/ops/retrieve/group_by_utils.cc [hunk 3]
```diff
@@ -38,10 +38,13 @@ struct GKey : public KeyBase {
     flat_hash_map<std::string_view, sel_t> sig_to_root;
     vector_t<vector_t<char>> root_list;
     for (size_t i = 0; i < row_num; ++i) {
-      vector_t<char> buf;
+      vector_t<char> buf((exprs.size() + 7) / 8, 0);
```
Comment (liulx20) [inline src/execution/execute/ops/retrieve/group_by_utils.cc]: 这里要不加个分支吧，先check一下exprs，是否会有optional，如果没有就无需这个bitset的开销

**Comment by liulx20** [comment]
Could we keep `Key` as the non-null typed fast path and let `GKey` be the fallback for nullable keys?

`create_key_func()` already has this separation. For a single mapping, it could check `chunk.get(...)->is_optional()` before calling `create_sp_key()`; optional columns would then fall through to `GKey`. The NULL-group handling in `Key::group()` could be removed.

This keeps the typed hot path free of a per-row NULL check and also avoids the current `has_value(i)` followed by `expr(i)` double read, which is especially relevant for vertex columns. If nullable single-key performance becomes important, a dedicated `NullableKey<T>` would be cleaner than mixing nullable behavior into `Key<T>`.

One prerequisite is that `GKey` must make NULL encoding collision-safe for every supported key shape. The bitmap in this patch only covers top-level expressions, so nested NULLs in LIST/ARRAY/STRUCT still need separate handling.

**Review by liulx20 — DISMISSED** [review]

File: src/execution/execute/ops/retrieve/group_by_utils.cc [hunk 4]
```diff
@@ -26,39 +26,67 @@ namespace ops {
...
-      } else {
-        sig_to_root.emplace(sv, groups.size());
-        root_list.emplace_back(std::move(buf));
-        offsets.push_back(i);
-        sel_vec_t ret_elem;
-        ret_elem.push_back(i);
-        groups.emplace_back(std::move(ret_elem));
+    } else {
+      for (size_t i = 0; i < row_num; ++i) {
+        vector_t<char> buf((exprs.size() + 7) / 8, 0);
```
Comment (longbinlai) [inline src/execution/execute/ops/retrieve/group_by_utils.cc:74]: group-by 层的修复本身是对的（bitmap 编码把 null 与任何值解耦，typed 路径经 `is_optional` 路由退回 GKey 我都核过），但 PR 写的 "Fixes #906" 与实测不符：**issue #906 的两条原始 repro 在本 PR 修复后依然是错的**，因为 null 在上游 UNWIND 求值阶段就被销毁了，根本到不了 group-by。

我在 pre-fix 构建上实证（直接 `RETURN x` 不带聚合即可看到）：

```text
UNWIND CAST([CAST(-1,'INT32'), CAST(NULL,'INT32')], 'INT32[]') AS x RETURN x
→ [[-1], [-1]]          # null 变成了前一个元素的副本

UNWIND CAST([0, CAST(NULL,'INT64')], 'INT64[]') AS x RETURN x
→ [[0], [0]]            # 同上
```

机制：`ValueColumnBuilder::push_back_elem`（value_columns.h:121）对 null Value 调 `GetValue<T>()`，后者（value.cc:492）不做 null 检查直接读 union，读到的是上一个 Value 的残留值。UNWIND 的两条路径（`unfold_list_like`/`unfold_impl`）都走这个函数，产出的列既不含 null 也不标 optional。

所以 case 1 里 null 变成 0 的副本、case 2 里变成 -1 的副本，issue 报的两条错误结果在上游就注定了，本 PR 改不到。建议：
1. PR 描述改为部分修复（group-by 路径），去掉 "Fixes #906" 的自动关闭，或补上 unfold 的 null 传递修复；
2. 另开 issue 跟踪 UNWIND null 毁坏（读 union 残留本质是读未定义数据——我探针跑批时同一路径还撞到过一次未复现的 SIGSEGV）。

File: tools/python_bind/tests/test_query.py [hunk 5]
```diff
@@ -82,6 +82,38 @@ def test_empty_ungrouped_count(empty_db):
     assert list(result) == [[0]]
 
 
+def test_group_by_preserves_null_keys(empty_db):
```
Comment (longbinlai) [inline tools/python_bind/tests/test_query.py:85]: 这两条断言没有回归能力：我把整个测试在 **pre-fix** 构建上跑过，两条全部通过（`[[-1, 1], [None, 2]]` / `[[-1, 7, 1], [None, 7, 2]]` 与期望一致）。原因：

1. 测试数据里没有 INT64 `0` 也没有 INT32 `-1`（PR body 声称 "covering NULL collisions with valid INT64 0 and INT32 -1 values"，与实际数据不符）——null 压成 0 后没有真实 0 跟它撞；target.id 是 INT64，null 的 `put_int(-1)` 与 INT64 -1 的 `put_long(-1)` 字节数不同，也不会撞。
2. 输出 key 列是从原列 reshuffle 的（group_by.cc:28-30），分组错了 key 显示照样正确，只有 COUNT 能暴露，而数据又构造不出碰撞。

最小改法：给 target 再插一行 `id: 0` 并连到某个 source，让 null（压成 0 后）与真实 0 同现——pre-fix 会输出 `[[0, N]]` 合并行、修复后输出 `[[0, 1], [None, N-1]]`，断言立刻有杀伤力。

Comment (Spockkk0225) [reply tools/python_bind/tests/test_query.py:85]: 这里确实有问题。在第一个case中，单列group by的NULL应该和0碰撞而不是-1；在第二个case中，多列group by的NULL应该和INT32的-1碰撞，需要修改类型。https://github.com/alibaba/neug/pull/907/commits/c6ef9adf9231fd54301ba0106d5327d4fade342c

File: src/execution/execute/ops/retrieve/group_by_utils.cc (hunk 4 above)
Comment (Spockkk0225) [reply src/execution/execute/ops/retrieve/group_by_utils.cc:74]: 这个问题此前已由 afe62d3a fix: preserve null elements in list and array columns (#882) 修复，相关逻辑确保 UNWIND 遇到 elem.IsNull() 时调用 push_back_null()。

**Review by longbinlai — APPROVED** [review]
LGTM
